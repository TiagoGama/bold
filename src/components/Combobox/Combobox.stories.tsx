import { boolean, number, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'
import { Combobox } from './Combobox'

const fruits = [
  { value: 1, label: 'Apple' },
  { value: 2, label: 'Avocado' },
  { value: 3, label: 'Banana' },
  { value: 4, label: 'Blueberry' },
  { value: 5, label: 'Coconut' },
  { value: 6, label: 'Grape' },
  { value: 7, label: 'Lemon' },
  { value: 8, label: 'Mango' },
  { value: 9, label: 'Melon' },
  { value: 10, label: 'Orange' },
  { value: 11, label: 'Peach' },
  { value: 12, label: 'Pear' },
]

storiesOf('Components|Combobox', module).add('default', () => (
  <Combobox<typeof fruits[0]>
    label='Fruit'
    name='fruit'
    items={fruits}
    error={text('error', '')}
    menuMinWidth={number('menuMinWidth (px)', undefined)}
    itemToString={(item) => item?.label}
    //   itemIsEqual={(a, b) => a.value === b.value}
    placeholder={text('placeholder', 'Select a value...')}
    //   multiple={boolean('multiple', false)}
    clearable={boolean('clearable', true)}
    disabled={boolean('disabled', false)}
    openOnFocus={boolean('openOnFocus', true)}
    //   loading={boolean('loading', false)}
    //   onChange={action('changed')}
    //   onBlur={action('blur')}
    //   onFocus={action('focus')}
  />
))
