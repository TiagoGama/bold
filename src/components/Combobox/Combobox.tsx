import { useCombobox } from 'downshift'
import matchSorter from 'match-sorter'
import React, { CSSProperties, useRef, useState } from 'react'
import { usePopper } from 'react-popper'
import { useLocale } from '../../i18n'
import { focusBoxShadow, Theme, useStyles } from '../../styles'
import { composeHandlers, composeRefs } from '../../util/react'
import { FormControl } from '../FormControl'
import { useFormControl, UseFormControlProps } from '../../hooks/useFormControl'
import { TextInput, TextInputProps } from '../TextField'
import { Spinner } from '../Spinner'

export interface ComboboxProps<T = string> extends Omit<TextInputProps, 'value' | 'onChange'>, UseFormControlProps {
  value?: T
  items: T[]
  itemToString(item: T): string
  openOnFocus: boolean
  loading: boolean
  menuMinWidth?: number
  filter?(items: T[], filter: string): T[]
  onChange?: (newValue: T) => void
  onFilterChange?: (newValue: string) => void
}

export function Combobox<T = string>(props: ComboboxProps<T>) {
  const {
    value,
    items,
    loading,
    itemToString,
    menuMinWidth,
    openOnFocus,
    onClear,
    onChange,
    onFocus,
    onFilterChange,
    filter = (items, filter) => matchSorter(items, filter, { keys: [itemToString] }),
    ...rest
  } = props

  const locale = useLocale()
  const { classes, css } = useStyles(createStyles)

  const [currentFilter, setCurrentFilter] = useState('')
  const visibleItems = filter(items, currentFilter)

  const inputRef = useRef<HTMLInputElement>()
  const [menuRef, setMenuRef] = useState<HTMLDivElement>()

  const {
    isOpen,
    highlightedIndex,
    getLabelProps,
    getMenuProps,
    getInputProps,
    getComboboxProps,
    getItemProps,
    openMenu,
    toggleMenu,
    closeMenu,
    reset,
  } = useCombobox<T>({
    selectedItem: value,
    items: visibleItems,
    itemToString,
    onInputValueChange: ({ inputValue }) => {
      setCurrentFilter(inputValue)
      onFilterChange?.(inputValue)
    },
    onSelectedItemChange: ({ selectedItem }) => {
      closeMenu()
      onChange?.(selectedItem)
    },
  })

  const downshiftComboboxProps = getComboboxProps()
  const { getFormControlProps, getInputProps: getFromControlInputProps } = useFormControl(props)
  const { ref: downshiftInputRef, ...downshiftInputProps } = getInputProps({
    onFocus: composeHandlers(onFocus, () => openOnFocus && openMenu()),
  })
  const { id: labelId, ...downshiftLabelProps } = getLabelProps()
  const downshiftMenuProps = getMenuProps()

  const {
    styles: { popper: popperStyles },
    attributes: { popper: popperAttributes },
  } = usePopper(inputRef.current, menuRef, {
    placement: 'bottom-start',
  })

  const formControlInputProps = getFromControlInputProps()
  const formControlProps = getFormControlProps()
  const invalid = !!formControlProps.error
  return (
    <div {...downshiftComboboxProps}>
      <FormControl {...formControlProps} labelId={labelId} {...downshiftLabelProps}>
        <TextInput
          icon={isOpen ? 'angleUp' : 'angleDown'}
          iconAriaLabel={isOpen ? locale.combobox.hideOptions : locale.combobox.showOptions}
          iconPosition='right'
          onIconClick={toggleMenu}
          inputRef={composeRefs(inputRef, downshiftInputRef)}
          onClear={composeHandlers(reset, onClear)}
          invalid={invalid}
          {...formControlInputProps}
          {...downshiftInputProps}
          {...rest}
        />
      </FormControl>

      {/*By the ARIA definition, the menu element should always be in the DOM*/}
      <div aria-busy={loading} {...downshiftMenuProps}>
        {isOpen && (
          <div
            data-testid='menu'
            className={classes.menu}
            style={{ ...popperStyles, width: inputRef.current?.clientWidth, minWidth: menuMinWidth }}
            {...popperAttributes}
            ref={setMenuRef}
          >
            <ul className={classes.list}>
              {loading && (
                <li className={css(classes.item, classes.loadingItem)}>
                  {locale.select.loadingItem}
                  <Spinner style={classes.loadingSpinner} />
                </li>
              )}
              {visibleItems.map((item, index) => (
                <li
                  className={css(classes.item, highlightedIndex === index && classes.selected)}
                  key={`${item}${index}`}
                  {...getItemProps({ item, index })}
                >
                  {itemToString(item)}
                </li>
              ))}
            </ul>
          </div>
        )}
      </div>
    </div>
  )
}

Combobox.defaultProps = {
  openOnFocus: true,
  loading: false,
} as Partial<ComboboxProps>

export const createStyles = (theme: Theme) => ({
  menu: {
    display: 'flex',
    flexDirection: 'column',
    zIndex: theme.zIndex.dropdown,
    border: `1px solid ${theme.pallete.divider}`,
    borderRadius: theme.radius.popper,
    backgroundColor: theme.pallete.surface.main,
    boxShadow: theme.shadows.outer['40'],
    maxHeight: '20rem',
  } as CSSProperties,

  list: {
    zIndex: 'auto',
    border: 0,
    borderRadius: 0,
    boxShadow: 'none',
    maxHeight: 'auto',
    listStyle: 'none',
    margin: 0,
    padding: 0,
    backgroundColor: theme.pallete.surface.main,
    overflowY: 'auto',
    overflowX: 'hidden',
    width: '100%',
  } as CSSProperties,

  item: {
    ...theme.typography.variant('main'),
    borderBottom: `1px solid ${theme.pallete.divider}`,
    cursor: 'pointer',
    padding: '0.5rem 0.5rem',
    transition: '.1s ease',

    '&:last-of-type': {
      borderBottom: 'none',
    },

    '&:hover': {
      background: theme.pallete.surface.background,
    },

    '&:focus': {
      outline: 0,
      borderRadius: 3,
      boxShadow: focusBoxShadow(theme, 'primary', 'inset'),
    },
  },

  loadingItem: {
    background: theme.pallete.surface.background,
    paddingTop: '0.25rem',
    paddingBottom: '0.25rem',
    cursor: 'initial',
  },

  loadingSpinner: {
    color: theme.pallete.primary.main,
    float: 'right',
  } as CSSProperties,

  selected: {
    outline: 0,
    background: theme.pallete.surface.background,
  },
})
