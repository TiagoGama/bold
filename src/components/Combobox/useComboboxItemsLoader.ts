import { useCallback, useMemo, useState } from 'react'

import { debounce } from '../../util'

export type AsyncComboboxLoadFn<T> = (filter: string) => Promise<T[]> | T[]

export function useComboboxItemsLoader<T>(items: AsyncComboboxLoadFn<T>, debounceMs = 350) {
  const [loading, setLoading] = useState(false)
  const [currItems, setCurrItems] = useState<T[]>([])

  const load = useCallback(
    (filter: string) => {
      setLoading(true)
      Promise.resolve(items(filter))
        .then(setCurrItems)
        .finally(() => setLoading(false))
    },
    [items]
  )

  const debouncedLoad = useMemo(() => debounce(load, debounceMs), [load, debounceMs])

  const loadItems = (filter: string) => {
    setLoading(true)
    return debouncedLoad(filter)
  }

  return {
    items: currItems,
    loading,
    loadItems,
  }
}
