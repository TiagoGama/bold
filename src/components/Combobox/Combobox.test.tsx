import React from 'react'
import { act, render, fireEvent, RenderResult, getByTestId } from '@testing-library/react'
import { Combobox, ComboboxProps } from './Combobox'

interface Fruit {
  value: number
  label: string
}

const fruits: Fruit[] = [
  { value: 1, label: 'Apple' },
  { value: 2, label: 'Avocado' },
  { value: 3, label: 'Banana' },
  { value: 4, label: 'Blueberry' },
  { value: 5, label: 'Coconut' },
  { value: 6, label: 'Grape' },
  { value: 7, label: 'Lemon' },
  { value: 8, label: 'Mango' },
  { value: 9, label: 'Melon' },
  { value: 10, label: 'Orange' },
  { value: 11, label: 'Peach' },
  { value: 12, label: 'Pear' },
]

const itemToString = (item: Fruit) => item.label

const ComboboxTest = (props: Partial<ComboboxProps<Fruit>>) => (
  <Combobox<typeof fruits[0]> items={fruits} itemToString={itemToString} {...props} />
)

it('has aria-compliant attributes', async () => {
  // From https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html
  let baseElement: RenderResult['baseElement']
  await act(async () => {
    const result = render(<ComboboxTest label='Fruits' />)
    baseElement = result.baseElement
  })
  const combobox = baseElement.querySelector('[role="combobox"]')
  const label = baseElement.querySelector('label')
  const input = baseElement.querySelector('input')
  const listbox = baseElement.querySelector('[role="listbox"]')
  const dropdownButton = baseElement.querySelector('button')

  expect(combobox).toHaveAttribute('aria-owns', listbox.getAttribute('id'))
  expect(combobox).toHaveAttribute('aria-expanded', 'false')
  expect(combobox).toHaveAttribute('aria-haspopup', 'listbox')

  expect(label).toHaveAttribute('id')
  expect(label).toHaveAttribute('for', input.getAttribute('id'))
  expect(label).toHaveTextContent('Fruits')

  expect(input).toHaveAttribute('id')
  expect(input).toHaveAttribute('aria-autocomplete', 'list')
  expect(input).toHaveAttribute('aria-controls', listbox.getAttribute('id'))
  expect(input).toHaveAttribute('aria-labelledby', label.getAttribute('id'))

  expect(dropdownButton).toHaveAttribute('tabindex', '-1')
  expect(dropdownButton).toHaveAttribute('aria-label')

  expect(listbox).toHaveAttribute('id')
  expect(listbox).toHaveAttribute('aria-labelledby', label.getAttribute('id'))

  await act(async () => {
    fireEvent.click(dropdownButton)
  })
  expect(combobox).toHaveAttribute('aria-expanded', 'true')
  expect(listbox.querySelector('[aria-selected]')).toBeTruthy()
})

it('opens menu when input is focused and only when `openOnFocus` prop is true', async () => {
  let baseElement: RenderResult['baseElement']
  let rerender: RenderResult['rerender']
  await act(async () => {
    const result = render(<ComboboxTest />)
    baseElement = result.baseElement
    rerender = result.rerender
  })
  const input = baseElement.querySelector('input')

  // initial state has closed menu
  expect(baseElement.querySelector('ul')).toBeFalsy()

  // focus input to open menu
  await act(async () => {
    fireEvent.focus(input)
  })
  expect(baseElement.querySelector('ul')).toBeTruthy()

  // blur input to close menu
  await act(async () => {
    fireEvent.blur(input)
  })
  expect(baseElement.querySelector('ul')).toBeFalsy()
  // rerenders switching prop
  await act(async () => rerender(<ComboboxTest openOnFocus={false} />))

  // focus input and now menu should not be open
  await act(async () => {
    fireEvent.focus(input)
  })
  expect(baseElement.querySelector('ul')).toBeFalsy()
})

it('shows placeholder when specified', async () => {
  let baseElement: RenderResult['baseElement']
  const placeholder = 'test placeholder'
  await act(async () => {
    const result = render(<ComboboxTest placeholder={placeholder} />)
    baseElement = result.baseElement
  })
  const input = baseElement.querySelector('input')

  expect(input).toHaveAttribute('placeholder', placeholder)
})

it('does not show placeholder when not specified', async () => {
  let baseElement: RenderResult['baseElement']
  await act(async () => {
    const result = render(<ComboboxTest />)
    baseElement = result.baseElement
  })
  const input = baseElement.querySelector('input')

  expect(input).not.toHaveAttribute('placeholder')
})

it('clears selection when "Clear" is clicked', async () => {
  let baseElement: RenderResult['baseElement']
  await act(async () => {
    const result = render(<ComboboxTest clearable={true} />)
    baseElement = result.baseElement
  })
  const input = baseElement.querySelector('input')

  const dropdownButton = baseElement.querySelector('button')
  //Opens menu
  await act(async () => {
    fireEvent.click(dropdownButton)
  })

  const option = baseElement.querySelector('li').firstChild

  //Selects item
  await act(async () => {
    fireEvent.click(option)
  })

  expect(input).toHaveValue(option.textContent)

  const clearButton = baseElement.querySelector('[title="Clear"]')

  //Clears value and focus out
  await act(async () => {
    fireEvent.click(clearButton)
  })
  await act(async () => {
    fireEvent.blur(input)
  })

  //Checks if cleared
  expect(input).not.toHaveValue()
})

it('enters error state', async () => {
  let baseElement: RenderResult['baseElement']
  const errorMessage = 'error'
  await act(async () => {
    const result = render(<ComboboxTest error={errorMessage} />)
    baseElement = result.baseElement
  })
  const input = baseElement.querySelector('input')

  expect(input).toHaveAttribute('aria-invalid', 'true')
  expect(input).toHaveAttribute('aria-errormessage')

  const errorMessageContainer = baseElement.querySelector(`#${input.getAttribute('aria-errormessage')}`)
  expect(errorMessageContainer).toHaveTextContent(errorMessage)
})

it('respects menu min-width', async () => {
  let baseElement: RenderResult['baseElement']

  await act(async () => {
    const result = render(<ComboboxTest menuMinWidth={1000} />)
    baseElement = result.baseElement
  })

  const dropdownButton = baseElement.querySelector('button')
  //Opens menu
  await act(async () => {
    fireEvent.click(dropdownButton)
  })

  const menu = getByTestId(baseElement, 'menu')

  expect(menu).toHaveStyle('min-width: 1000px')
})

it('should trigger onChange', async () => {
  let baseElement: RenderResult['baseElement']

  let selection = null

  await act(async () => {
    const result = render(<ComboboxTest onChange={(nValue) => (selection = nValue)} />)
    baseElement = result.baseElement
  })

  expect(selection).toBeNull()

  //Opens menu
  const dropdownButton = baseElement.querySelector('button')
  await act(async () => {
    fireEvent.click(dropdownButton)
  })

  //Selects first item
  const option = baseElement.querySelector('li').firstChild
  await act(async () => {
    fireEvent.click(option)
  })

  expect(selection).toBe(fruits[0])
})

it('should trigger onFilterChange', async () => {
  let baseElement: RenderResult['baseElement']

  let filter = ''

  await act(async () => {
    const result = render(<ComboboxTest onFilterChange={(nValue) => (filter = nValue)} />)
    baseElement = result.baseElement
  })

  expect(filter).toBe('')

  const input = baseElement.querySelector('input')
  fireEvent.change(input, { target: { value: 'filter' } })

  expect(filter).toBe('filter')
})

it('should trigger onFocus', async () => {
  let baseElement: RenderResult['baseElement']

  const onFocus = jest.fn()

  await act(async () => {
    const result = render(<ComboboxTest onFocus={onFocus} />)
    baseElement = result.baseElement
  })

  const input = baseElement.querySelector('input')
  fireEvent.focus(input)
  expect(onFocus).toBeCalled()
})

it('should trigger onBlur', async () => {
  let baseElement: RenderResult['baseElement']

  const onBlur = jest.fn()

  await act(async () => {
    const result = render(<ComboboxTest onBlur={onBlur} />)
    baseElement = result.baseElement
  })

  const input = baseElement.querySelector('input')
  fireEvent.focus(input)
  fireEvent.blur(input)
  expect(onBlur).toBeCalled()
})

it('should accept a value as parameter', async () => {
  let baseElement: RenderResult['baseElement']

  await act(async () => {
    const result = render(<ComboboxTest value={fruits[1]} />)
    baseElement = result.baseElement
  })

  const input = baseElement.querySelector('input')

  expect(input).toHaveValue(itemToString(fruits[1]))

  //Opens menu
  const dropdownButton = baseElement.querySelector('button')
  await act(async () => {
    fireEvent.click(dropdownButton)
  })

  const listbox = baseElement.querySelector('[role="listbox"]')
  const selected = listbox.querySelector('[aria-selected="true"]')

  expect(selected).toHaveTextContent(itemToString(fruits[1]))
})

it('renders correcly closed', async () => {
  let baseElement: RenderResult['baseElement']
  await act(async () => {
    const result = render(<ComboboxTest label='Fruits' />)
    baseElement = result.baseElement
  })
  expect(baseElement).toMatchSnapshot()
})

it('renders correcly opened', async () => {
  let baseElement: RenderResult['baseElement']
  await act(async () => {
    const result = render(<ComboboxTest label='Fruits' />)
    baseElement = result.baseElement
  })
  const dropdownButton = baseElement.querySelector('button')
  await act(async () => {
    fireEvent.click(dropdownButton)
  })
  expect(baseElement).toMatchSnapshot()
})
