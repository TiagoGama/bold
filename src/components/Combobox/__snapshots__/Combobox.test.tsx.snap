// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correcly closed 1`] = `
.emotion-0 {
  font-weight: bold;
  display: block;
  margin-bottom: 0.25rem;
  line-height: 20px;
}

.emotion-5 {
  position: relative;
}

.emotion-5 input {
  padding-right: 3rem;
}

.emotion-4 {
  position: absolute;
  background-color: #F0F0F5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.25rem;
  width: 2.5rem;
  top: 1px;
  bottom: 1px;
  right: 1px;
}

.emotion-1 {
  background-color: #FFFFFF;
  border: solid 1px #8F8FA2;
  border-radius: 2px;
  color: #24252E;
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1rem;
  margin: 0;
  padding: calc(0.5rem - 1px) 0.5rem;
  width: 100%;
  -webkit-transition-property: box-shadow;
  transition-property: box-shadow;
  -webkit-transition-duration: .2s;
  transition-duration: .2s;
}

.emotion-1[type="search"]::-webkit-search-decoration {
  display: none;
}

.emotion-1[type="search"]::-webkit-search-cancel-button {
  display: none;
}

.emotion-1[type="search"]::-webkit-search-results-button {
  display: none;
}

.emotion-1[type="search"]::-webkit-search-results-decoration {
  display: none;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1::-webkit-input-placeholder {
  color: #696979;
}

.emotion-1::-moz-placeholder {
  color: #696979;
}

.emotion-1:-ms-input-placeholder {
  color: #696979;
}

.emotion-1::placeholder {
  color: #696979;
}

.emotion-1:disabled {
  border-color: #D3D4DD;
  background-color: #F0F0F5;
}

.emotion-1:not(:disabled):hover {
  border-color: #7A7B8E;
}

.emotion-1:not(:disabled):focus {
  outline: none;
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-1:not(:disabled):active {
  border-color: #0069D0;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.09);
}

.emotion-3 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5rem;
  color: #24252E;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  font-weight: bold;
  cursor: pointer;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 4px;
  color: #24252E;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 0.875rem;
  padding: calc(0.25rem - 1px) calc(0.5rem - 1px);
  border-radius: inherit;
}

.emotion-3 > span {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-3:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,0.1);
}

.emotion-3:focus {
  outline: none;
  box-shadow: 0 0 0 2px #0069D0;
}

.emotion-3:not(:disabled):hover {
  background-color: rgba(58,58,71,0.1);
}

.emotion-3:focus {
  box-shadow: none;
}

.emotion-2 {
  fill: currentColor;
  font-size: 1.5rem;
}

<body>
  <div>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-21-menu"
      role="combobox"
    >
      <div
        class="emotion-6"
      >
        <label
          class="emotion-0"
          for="downshift-20-input"
          id="downshift-21-label"
        >
          Fruits
        </label>
        <div
          class="emotion-5"
        >
          <input
            aria-autocomplete="list"
            aria-controls="downshift-21-menu"
            aria-labelledby="downshift-21-label"
            autocomplete="off"
            class="emotion-1"
            id="downshift-20-input"
            type="text"
            value=""
          />
          <span
            class="emotion-4"
          >
            <button
              aria-label="Show options"
              class="emotion-3"
              tabindex="-1"
              type="button"
            >
              <span>
                <svg
                  aria-hidden="true"
                  class="emotion-2"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                >
                  <path
                    clip-rule="evenodd"
                    d="M16.294 8.294a1 1 0 111.414 1.414l-5 5a1 1 0 01-1.414 0l-5.001-5a1 1 0 011.414-1.415l4.294 4.294 4.293-4.293z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </div>
      </div>
      <div
        aria-labelledby="downshift-21-label"
        id="downshift-21-menu"
        role="listbox"
      />
    </div>
  </div>
</body>
`;

exports[`renders correcly opened 1`] = `
.emotion-20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  z-index: 1300;
  border: 1px solid #D3D4DD;
  border-radius: 2px;
  background-color: #FFFFFF;
  box-shadow: 0 2px 4px -1px rgba(0,0,0,0.2),0 1px 10px 0 rgba(0,0,0,0.12),0 4px 5px 0 rgba(0,0,0,0.14);
  max-height: 20rem;
}

.emotion-19 {
  z-index: auto;
  border: 0;
  border-radius: 0;
  box-shadow: none;
  max-height: auto;
  list-style: none;
  margin: 0;
  padding: 0;
  background-color: #FFFFFF;
  overflow-y: auto;
  overflow-x: hidden;
  width: 100%;
}

.emotion-7 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5;
  color: #24252E;
  border-bottom: 1px solid #D3D4DD;
  cursor: pointer;
  padding: 0.5rem 0.5rem;
  -webkit-transition: .1s ease;
  transition: .1s ease;
}

.emotion-7:last-of-type {
  border-bottom: none;
}

.emotion-7:hover {
  background: #F0F0F5;
}

.emotion-7:focus {
  outline: 0;
  border-radius: 3px;
  box-shadow: inset 0 0 0 2px #0069D0;
}

.emotion-0 {
  font-weight: bold;
  display: block;
  margin-bottom: 0.25rem;
  line-height: 20px;
}

.emotion-5 {
  position: relative;
}

.emotion-5 input {
  padding-right: 3rem;
}

.emotion-4 {
  position: absolute;
  background-color: #F0F0F5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.25rem;
  width: 2.5rem;
  top: 1px;
  bottom: 1px;
  right: 1px;
}

.emotion-1 {
  background-color: #FFFFFF;
  border: solid 1px #8F8FA2;
  border-radius: 2px;
  color: #24252E;
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1rem;
  margin: 0;
  padding: calc(0.5rem - 1px) 0.5rem;
  width: 100%;
  -webkit-transition-property: box-shadow;
  transition-property: box-shadow;
  -webkit-transition-duration: .2s;
  transition-duration: .2s;
}

.emotion-1[type="search"]::-webkit-search-decoration {
  display: none;
}

.emotion-1[type="search"]::-webkit-search-cancel-button {
  display: none;
}

.emotion-1[type="search"]::-webkit-search-results-button {
  display: none;
}

.emotion-1[type="search"]::-webkit-search-results-decoration {
  display: none;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1::-webkit-input-placeholder {
  color: #696979;
}

.emotion-1::-moz-placeholder {
  color: #696979;
}

.emotion-1:-ms-input-placeholder {
  color: #696979;
}

.emotion-1::placeholder {
  color: #696979;
}

.emotion-1:disabled {
  border-color: #D3D4DD;
  background-color: #F0F0F5;
}

.emotion-1:not(:disabled):hover {
  border-color: #7A7B8E;
}

.emotion-1:not(:disabled):focus {
  outline: none;
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-1:not(:disabled):active {
  border-color: #0069D0;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.09);
}

.emotion-3 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5rem;
  color: #24252E;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  font-weight: bold;
  cursor: pointer;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 4px;
  color: #24252E;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 0.875rem;
  padding: calc(0.25rem - 1px) calc(0.5rem - 1px);
  border-radius: inherit;
}

.emotion-3 > span {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-3:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,0.1);
}

.emotion-3:focus {
  outline: none;
  box-shadow: 0 0 0 2px #0069D0;
}

.emotion-3:not(:disabled):hover {
  background-color: rgba(58,58,71,0.1);
}

.emotion-3:focus {
  box-shadow: none;
}

.emotion-2 {
  fill: currentColor;
  font-size: 1.5rem;
}

<body>
  <div>
    <div
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-owns="downshift-23-menu"
      role="combobox"
    >
      <div
        class="emotion-6"
      >
        <label
          class="emotion-0"
          for="downshift-22-input"
          id="downshift-23-label"
        >
          Fruits
        </label>
        <div
          class="emotion-5"
        >
          <input
            aria-autocomplete="list"
            aria-controls="downshift-23-menu"
            aria-labelledby="downshift-23-label"
            autocomplete="off"
            class="emotion-1"
            id="downshift-22-input"
            type="text"
            value=""
          />
          <span
            class="emotion-4"
          >
            <button
              aria-label="Hide options"
              class="emotion-3"
              tabindex="-1"
              type="button"
            >
              <span>
                <svg
                  aria-hidden="true"
                  class="emotion-2"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                >
                  <path
                    clip-rule="evenodd"
                    d="M7.707 14.707a1 1 0 11-1.414-1.414l5-5a1 1 0 011.414 0l5.001 5.001a1 1 0 01-1.414 1.414L12 10.414l-4.293 4.293z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </div>
      </div>
      <div
        aria-labelledby="downshift-23-label"
        id="downshift-23-menu"
        role="listbox"
      >
        <div
          class="emotion-20"
          style="position: absolute; left: 0px; top: 0px; width: 0px;"
        >
          <ul
            class="emotion-19"
          >
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-0"
              role="option"
            >
              Apple
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-1"
              role="option"
            >
              Avocado
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-2"
              role="option"
            >
              Banana
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-3"
              role="option"
            >
              Blueberry
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-4"
              role="option"
            >
              Coconut
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-5"
              role="option"
            >
              Grape
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-6"
              role="option"
            >
              Lemon
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-7"
              role="option"
            >
              Mango
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-8"
              role="option"
            >
              Melon
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-9"
              role="option"
            >
              Orange
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-10"
              role="option"
            >
              Peach
            </li>
            <li
              aria-selected="false"
              class="emotion-7"
              id="downshift-23-item-11"
              role="option"
            >
              Pear
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</body>
`;
